### PR TITLE
feat: eden dataloader

### PR DESCRIPTION
**Pull request by Geraldene**
### Type of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactor
- [ ] Documentation update
- [ ] Other (please describe):

### CI Pipeline Configuration

Configure CI behavior by applying the relevant labels:

- [SKIP_CI](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#skip_ci) - Skip all continuous integration tests
- [INCLUDE_NOTEBOOKS_TESTS](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#include_notebooks_tests) - Execute notebook validation tests in pytest
- [INCLUDE_SLOW_TESTS](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/user-guide/contributing/contributing.md#include_slow_tests) - Execute tests labelled as slow in pytest for extensive testing

> [!NOTE]
> By default, the notebooks validation tests are skipped unless explicitly enabled.

#### Authorizing CI Runs

We use [copy-pr-bot](https://docs.gha-runners.nvidia.com/apps/copy-pr-bot/#automation) to manage authorization of CI
runs on NVIDIA's compute resources.

- If a pull request is opened by a trusted user and contains only trusted changes, the pull request's code will
  automatically be copied to a pull-request/ prefixed branch in the source repository (e.g. pull-request/123)
- If a pull request is opened by an untrusted user or contains untrusted changes, an NVIDIA org member must leave an
  `/ok to test` comment on the pull request to trigger CI. This will need to be done for each new commit.

### Usage

```python
# TODO: Add code snippet
```

### Pre-submit Checklist

- [ ] I have tested these changes locally
- [ ] I have updated the documentation accordingly
- [ ] I have added/updated tests as needed
- [ ] All existing tests pass successfully

## Summary by CodeRabbit

* **New Features**
  * High-performance sharded DataLoader with precomputed window mappings, per-sample SQLite-backed sequence stores, CLI precompute tool, and optional run-local window access logging.
  * Expanded multi-model training support (HYENA / MAMBA / LLAMA) and new Eden/LLAMA configuration variants.

* **Refactor**
  * Training flow unified around evo2_config with clearer dataset sources, deterministic tokenizer initialization, and updated checkpoint/save behavior.

* **Documentation**
  * Added user guide and examples for the sharded DataLoader and preprocessing workflow.

### CODE REVIEW

File: sub-packages/bionemo-evo2/src/bionemo/evo2/models/llama.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+    "8B": lambda **kwargs: llm.Llama3Config8B(**kwargs),
+    "7B": lambda **kwargs: EdenConfig(**kwargs),
+    "11B": lambda **kwargs: Eden11BConfig(**kwargs),
+    "18B": lambda **kwargs: Eden18BConfig(**kwargs),
+    "21B": lambda **kwargs: Eden21BConfig(**kwargs),
+    "24B": lambda **kwargs: Eden24BConfig(**kwargs),
+    "27B": lambda **kwargs: Eden27BConfig(**kwargs),
+    "28B": lambda **kwargs: Eden28BConfig(**kwargs),
+    "35B": lambda **kwargs: Eden35BConfig(**kwargs),
+}
```
Comment (jstjohn) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/models/llama.py:203]: Why are the lambda's needed? Can't you just do:
`"8B": llm.Llama3Config8B,` which points to the uninitialize class type? Then the user can call that with `LLAMA_MODEL_OPTIONS["8b"](**kwargs)` as many times as they want still to get different initialized versions of the class.

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py [hunk 2]
```diff
@@ -443,12 +672,6 @@ def parse_args(args: Optional[List[str]] = None) -> argparse.Namespace:
         default=0.0,
         help="Dropout probability for the hyena layers",
     )
-    parser.add_argument(
-        "--ffn-hidden-size",
-        type=int,
-        default=None,
-        help="FFN hidden size for the hyena layers",
-    )
```
Comment (jstjohn) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py:451]: Can you add this back in?

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py [hunk 3]
```diff
@@ -383,13 +619,6 @@ def parse_args(args: Optional[List[str]] = None) -> argparse.Namespace:
         default=False,
         help="Do not renormalize the loss weights.",
     )
-    parser.add_argument(
-        "--mamba-lowercase-loss-weight",
-        type=float,
-        default=0.1,
-        help="Loss weight for the Mamba model for lowercase bases, if you are using a Mamba model. "
-        "Default is 0.1 like the Evo2 paper. Set to 1.0 to disable differential loss weighting.",
-    )
```
Comment (jstjohn) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py:392]: Add this back in as well please.

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py [hunk 4]
```diff
@@ -462,9 +685,9 @@ def parse_args(args: Optional[List[str]] = None) -> argparse.Namespace:
         help="Dropout probability for the attention layers.",
     )
     parser.add_argument(
-        "--use-subquadratic_ops",
+        "--use-b2b-causal-conv1d",
```
Comment (jstjohn) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py:688]: Please revert the names here to subquadratic ops

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py [hunk 5]
```diff
@@ -462,9 +685,9 @@ def parse_args(args: Optional[List[str]] = None) -> argparse.Namespace:
...
         action="store_true",
-        help="Use subquadratic_ops for improved performance.",
+        help="Use back-to-back causal convolution CUDA kernel for hyena short conv layers for improved performance.",
```
Comment (jstjohn) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py:690]: Same here

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py [hunk 6]
```diff
@@ -491,33 +714,28 @@ def parse_args(args: Optional[List[str]] = None) -> argparse.Namespace:
...
         help="Disable saving the last checkpoint.",
     )
-    parser.add_argument("--lora-finetune", action="store_true", help="Use LoRA fine-tuning", default=False)
-    parser.add_argument("--lora-checkpoint-path", type=Path, default=None, help="LoRA checkpoint path")
-    parser.add_argument(
-        "--no-calculate-per-token-loss",
-        action="store_true",
-        default=False,
-        help="Calculate a simpler mean across the microbatch of the loss prior to DDP reduction rather than the global"
-        " per-token mean loss. Use this if speed is critical and if you do not need token masking in your loss.",
```
Comment (jstjohn) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py:501]: Please add these options back

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py [hunk 7]
```diff
@@ -575,30 +835,20 @@ def train(args: argparse.Namespace) -> nl.Trainer:
             }
         else:
             activation_checkpointing_args = {}
+
     # Retrieve model config.
     config_modifiers_init = {
-        "calculate_per_token_loss": not args.no_calculate_per_token_loss,  # override megatron internal behavior.
```
Comment (jstjohn) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py:580]: let's probably add this back, but I'm curious if this change hurts your training loss curve...

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py [hunk 8]
```diff
@@ -607,35 +857,40 @@ def train(args: argparse.Namespace) -> nl.Trainer:
...
             raise ValueError(f"Invalid model size for Mamba: {args.model_size}")
         add_bias_output = config_modifiers_init.pop("add_bias_output")
         if add_bias_output:
             raise ValueError("Bias output is not supported for Mamba models.")
+        config_modifiers_init.pop("use_b2b_causal_conv1d", None)
         model_config = MAMBA_MODEL_OPTIONS[args.model_size](**config_modifiers_init)
         model = MambaModel(model_config, tokenizer=data_module.tokenizer)
+    elif model_type == "llama":
+        if args.spike_no_more_embedding_init:  # --spike-no-more-embedding-init
+            config_modifiers_init["embedding_init_method_std"] = 1.0
```
Comment (jstjohn) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py:887]: This can move back to the global scope. It's a megatron setting so it will be supported in all three of these models.

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py [hunk 9]
```diff
@@ -575,30 +835,20 @@ def train(args: argparse.Namespace) -> nl.Trainer:
...
         "fp32_residual_connection": not args.no_fp32_residual_connection,
-        "add_bias_output": args.add_bias_output,
         **activation_checkpointing_args,
     }
-    if args.spike_no_more_embedding_init:
-        config_modifiers_init["embedding_init_method_std"] = 1.0
-        # When using spike_no_more_embedding_init, we don't want to share embeddings and outputs.
-        config_modifiers_init["share_embeddings_and_output_weights"] = False
-    if args.ffn_hidden_size:
-        config_modifiers_init["ffn_hidden_size"] = args.ffn_hidden_size
```
Comment (jstjohn) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py:597]: Please add these back.

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py [hunk 10]
```diff
@@ -575,30 +835,20 @@ def train(args: argparse.Namespace) -> nl.Trainer:
...
             activation_checkpointing_args = {}
+
     # Retrieve model config.
     config_modifiers_init = {
-        "calculate_per_token_loss": not args.no_calculate_per_token_loss,  # override megatron internal behavior.
         "tp_comm_overlap": args.use_megatron_comm_overlap_llama3_8k,
         "seq_length": args.seq_length,
         "hidden_dropout": args.hidden_dropout,
         "attention_dropout": args.attention_dropout,
-        "to_upper": "weighted" if args.no_renormalize_loss else "normalized_weighted",
```
Comment (jstjohn) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py:585]: This can move into the mamba/hyena configs, or you could pop it out of the llama config?

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py [hunk 11]
```diff
@@ -575,30 +835,20 @@ def train(args: argparse.Namespace) -> nl.Trainer:
...
         "distribute_saved_activations": False if args.sequence_parallel else True,
         "cross_entropy_loss_fusion": args.cross_entropy_loss_fusion,
         "fp32_residual_connection": not args.no_fp32_residual_connection,
-        "add_bias_output": args.add_bias_output,
```
Comment (jstjohn) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py:589]: Move this into a block like:

```
if args.add_bias_oputput:
    config_modifiers_init["add_bias_output"] = args.add_bias_oputput
```

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py [hunk 12]
```diff
@@ -575,30 +835,20 @@ def train(args: argparse.Namespace) -> nl.Trainer:
...
-    if args.spike_no_more_embedding_init:
-        config_modifiers_init["embedding_init_method_std"] = 1.0
-        # When using spike_no_more_embedding_init, we don't want to share embeddings and outputs.
-        config_modifiers_init["share_embeddings_and_output_weights"] = False
-    if args.ffn_hidden_size:
-        config_modifiers_init["ffn_hidden_size"] = args.ffn_hidden_size
     if args.use_targeted_variance_loss:
         config_modifiers_init["use_targeted_variance_loss"] = True
-    if args.use_subquadratic_ops:
-        config_modifiers_init["use_subquadratic_ops"] = True
```
Comment (jstjohn) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py:601]: Add this one back

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py [hunk 13]
```diff
@@ -251,13 +469,20 @@ def parse_args(args: Optional[List[str]] = None) -> argparse.Namespace:
         help="Restore optimizer state from initial checkpoint. Defaults to False.",
     )
     parser.add_argument(
-        "--average-in-collective",
+        "--no-average-in-collective",
```
Comment (jstjohn) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py:472]: I'd go back to making it `--average-in-collective` and make sure that the logic below on how that's set matches the argument.

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py [hunk 14]
```diff
@@ -607,35 +857,40 @@ def train(args: argparse.Namespace) -> nl.Trainer:
...
-        model = llm.HyenaModel(model_config, tokenizer=data_module.tokenizer, model_transform=lora_transform)
-    else:  # mamba
-        if args.no_weight_decay_embeddings:
-            config_modifiers_init["hyena_no_weight_decay_cond_fn"] = mamba_no_weight_decay_cond_with_embeddings
-        config_modifiers_init["lowercase_loss_reweighting"] = args.mamba_lowercase_loss_weight
+            config_modifiers_init["hyena_no_weight_decay_cond_fn"] = (
+                mamba_no_weight_decay_cond_with_embeddings
+            )
+        if args.spike_no_more_embedding_init:  # --spike-no-more-embedding-init
+            config_modifiers_init["spike_no_more_embedding_init"] = True
```
Comment (jstjohn) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py:876]: I think you can undo this change once you have the spike no more logic moved out to the main override like it was before. This also will get the lora support back for the evo2 hyena model.

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py [hunk 15]
```diff
@@ -607,35 +857,40 @@ def train(args: argparse.Namespace) -> nl.Trainer:
...
         if args.model_size not in MAMBA_MODEL_OPTIONS:
             raise ValueError(f"Invalid model size for Mamba: {args.model_size}")
         add_bias_output = config_modifiers_init.pop("add_bias_output")
         if add_bias_output:
             raise ValueError("Bias output is not supported for Mamba models.")
+        config_modifiers_init.pop("use_b2b_causal_conv1d", None)
```
Comment (jstjohn) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py:882]: Once you make the other changes, you'll want to pop the `use_subquadratic_ops` instead, although that's only added if asked for, so this would just protect against a user who says `--use-subquadratic-ops` and is also requesting a non-hyena model.

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py [hunk 16]
```diff
@@ -645,12 +900,12 @@ def train(args: argparse.Namespace) -> nl.Trainer:
         TEVCallback(),
     ]
 
-    if args.lora_finetune:
-        callbacks.append(ModelTransform())
```
Comment (jstjohn) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py:649]: Add this one back too.

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py [hunk 17]
```diff
@@ -758,12 +1037,26 @@ def train(args: argparse.Namespace) -> nl.Trainer:
         )
     )
     nemo_logger = setup_nemo_lightning_logger(
-        root_dir=args.result_dir,
+        root_dir=args.results_dir,
```
Comment (jstjohn) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py:1040]: `result-dir` is more consistent with other training scripts, so we made that change, can you change it back to result_dir?

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py [hunk 18]
```diff
@@ -799,12 +1092,12 @@ def train(args: argparse.Namespace) -> nl.Trainer:
         auto_resume = None
 
     ddp: DistributedDataParallelConfig = DistributedDataParallelConfig(
-        check_for_nan_in_grad=not args.no_check_for_nan_in_grad,
+        check_for_nan_in_grad=True,
```
Comment (jstjohn) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py:1095]: This is useful for debugging, can you add it back?

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py [hunk 19]
```diff
@@ -799,12 +1092,12 @@ def train(args: argparse.Namespace) -> nl.Trainer:
...
         overlap_grad_reduce=args.overlap_grad_reduce,
         overlap_param_gather=args.overlap_param_gather,  # Verify that this works using
         grad_reduce_in_fp32=args.grad_reduce_in_fp32,
         align_param_gather=args.align_param_gather,
-        average_in_collective=average_in_collective,
+        average_in_collective=not args.no_average_in_collective,
```
Comment (jstjohn) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py:1100]: This is related to the other change where we use per-token-loss by default. Per-token loss is supposedly better when you have padding and/or masked tokens in your loss. I'm curious if this detrimentally impacts your training though. Average in collective is not compatible with per-token loss so that's why the default changed from doing it to not doing it.

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py [hunk 20]
```diff
@@ -879,7 +1176,14 @@ def train(args: argparse.Namespace) -> nl.Trainer:
...
     # This is where the no weight decay condition is applied to the optimizer state.
-    opt = MegatronOptimizerModule(opt_config, sched, no_weight_decay_cond=model_config.hyena_no_weight_decay_cond_fn)
+    if (model_type == "hyena") | (model_type == "mamba"):
+        opt = MegatronOptimizerModule(
+            opt_config,
+            sched,
+            no_weight_decay_cond=model_config.hyena_no_weight_decay_cond_fn,
+        )
+    else:
+        opt = MegatronOptimizerModule(opt_config, sched)
```
Comment (jstjohn) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py:1186]: The consequence of this is you have no way to skip weight decay on your embedding layer. That may or may not matter, but it's interesting that you are doing weight decay on embeddings.

**Comment by yzhang123** [comment]
moved to https://github.com/NVIDIA/bionemo-framework/pull/1109

File: sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py [hunk 21]
```diff
@@ -539,11 +757,52 @@ def train(args: argparse.Namespace) -> nl.Trainer:
             num_workers=args.workers,
             tokenizer=tokenizer,
         )
+    elif args.fasta_data:
+        data_module = EdenDataModule(
```
Comment (yzhang123) [inline sub-packages/bionemo-evo2/src/bionemo/evo2/run/train.py:761]: Hi @Geraldene could it be you missed to add EdenDataModule source code?
